src/Stratis.Bitcoin/Utilities/AsyncConsumerQueue.cs

```diff
@@ -0,0 +1,154 @@
+
+        /// <summary>Task that waits until specified delay runs out.</summary>
+        private Task timerAwaiterTask;
+
+        /// <summary>Task that runs execution flow.</summary>
+        private Task executor;
+
+        /// <summary>Cancellation token source.</summary>
+        private CancellationTokenSource cancellation;
+
+        private object lockObject;
+
+        /// <summary>Queue items count.</summary>
+        public int Count => this.queue.Count;
+
+        /// <summary>
+        /// Creates new instance of <see cref="AsyncConsumerQueue{T}<T>"/>.
+        /// </summary>
+        /// <param name="processingAction">Action that consumes items from the queue and removes them when they are consumed.</param>
+        /// <param name="immediateProcessingTrigger">Condition that has to be met in order to start consumption action immediately.</param>
+        /// <param name="timer">Maximum amount of time that can pass since adding item to the queue before it gets consumed.</param>
+        /// <param name="restartTimerIfNotAllItemsConsumed">If set to <c>true</c> timer will be automatically restarted if not all items were consumed.</param>
+        public AsyncConsumerQueue(
+            Action<ConcurrentQueue<T>> processingAction,
+            Func<T, bool> immediateProcessingTrigger = null,
+            TimeSpan? timer = null,
+            bool restartTimerIfNotAllItemsConsumed = true)
+        {
+            this.queue = new ConcurrentQueue<T>();
+            this.processingAction = processingAction;
+            this.immediateProcessingTrigger = immediateProcessingTrigger;
+            this.timer = timer;
+            this.restartTimerIfNotAllItemsConsumed = restartTimerIfNotAllItemsConsumed;
+
+            this.trigger = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
+            this.lockObject = new object();
+        }
+
+        /// <inheritdoc />
+        public void Enqueue(T item)
+        {
+            this.queue.Enqueue(item);
+
+            Task.Run(() =>
+            {
+                if (this.immediateProcessingTrigger != null && this.immediateProcessingTrigger(item))
+                {
+                    this.trigger.Set();
+                }
+                else
+                {
+                    lock (this.lockObject)
+                    {
+                        if (this.timer != null && (this.timerAwaiterTask == null))
+                            SetTimer();
+                    }
+                }
+
+                lock (this.lockObject)
+                {
+                    // Ensure flow is running.
+                    if (this.executor == null || this.executor.IsCompleted)
+                        this.executor = ExecutionFlowAsync();
+                }
+
+            }).ConfigureAwait(false);
+        }
+
+        private void SetTimer()
+        {
+            this.timerAwaiterTask = Task.Delay(this.timer.Value, this.cancellation.Token);
```

Comment:
why would you not use Timer object, which is resetable, while cancellation is not
it seems that you again misuse cancellation for something it is not designed for
you don't need timer task at all imho

Reply:
Timer is not awaitable.

Reply:
why does it matter?